### ISSUE TITLE
How to connect hub to persistent storage?

### ISSUE THREAD

**Issue by dfelikson** [CryoInTheCloud/hub-image#30]
Our group wants to use the hub to collaborate on our ICESat-2 research project. We have several people across multiple institutions who are already using Jupyter notebooks for analysis and there's a lot of overlap in the data that everyone is using, including data products that are stored on DAACs as well as "intermediate" data products that are being derived from the DAAC datasets. Ideally, what we need is common cloud storage that we can share so that we all have access to these shared datasets.

Is there a way to connect the hub to persistent cloud storage? For example, could we set up an S3 bucket for our team to use that can be directly accessed by the hub?

Right now, we do a lot of file transfer back and forth between team members and we're storing the same data in multiple (local) places. Consolidating this data would make things a lot more efficient for us!

**Comment by yuvipanda** [comment]
So, hubs maintained by 2i2c for research generally have two object storage env vars set: https://infrastructure.2i2c.org/en/latest/topic/features.html#scratch-buckets-on-object-storage

`SCRATCH_BUCKET` is a s3 bucket with read/write permissions to all users, but lifecycle set to delete after every 7 days. This helps with intermediate products, and makes sharing easy - while `SCRATCH_BUCKET` env var for each user will have their username interpolated into it, users can access each others'. So you can write into your scratch bucket and someone else can read from it. It's also possible to arbitrarily create project specific ones - there's basically no access control at the path level :) 

There's also `PERSISTENT_BUCKET` which works exactly the same as `SCRATCH_BUCKET` but without the 7 day wipe rule. This allows for longer term sharing of intermediate products. *HOWEVER*, this can also baloon costs so someone needs to keep an eye on it.

I see this is currently not enabled on the cryo hub. Would you like us to enable it?

**Comment by dfelikson** [comment]
Thanks for everyone's responses so far. To answer @weiji14, I'll gather some better estimates from my group but we would be sharing data files that are mostly on the order of 100 MB, with some that are as large as 10 GB. It's possible we'd have some that are larger than that but not many. Again, I can get a better estimate next week. And we're looking at about 6-8 users who would be reading this data directly from the S3 storage.

As far as syncthing, I don't think it's the way for us to go for two reasons. One is that, as @weiji14 said, we'd be storing duplicate copies of data on everyone's home directories on the DaskHub (if I understand how it works correctly). The second is that I'd prefer not to make everyone sync every time a dataset was updated, although we can certainly live with this. The third is that I think we'd fill up everyone's home directory storage pretty quickly with the data.

We would definitely want this data to persist for the life of the project so `PERSISTENT_BUCKET` would be what's needed. @yuvipanda - I totally understand the issue of balooning cost. Maybe the next step is for me to get a better total estimate and we can see how much that would cost?

**Comment by yuvipanda** [comment]
@dfelikson yep that sounds like a great next step!

**Comment by dfelikson** [comment]
Hi all - I wanted to give an update on this. My team decided to purchase AWS S3 storage that's dedicated to our project (funded by NASA ROSES). I created a new bucket in `us-west-2` and created access keys for team members to be able to access the bucket. I'm in the process of writing a new entry for the CryoInTheCloud JupyterBook to describe the process of setting this up.

**Comment by dfelikson** [comment]
@tsnow03 - my team has been able to set up a project-specific AWS S3 bucket, and set the user policy so that the team can access it from CryoHub. I can write up a short JupyterBook entry for how to do this but I'm still having trouble setting up the conda environment. I'm using `conda env create -f conda/environment.yml` and trying on my laptop and on the CyoHub.

On my laptop, I get this error:
```
Solving environment: failed

ResolvePackageNotFound:
  - qgis==3.18.3
  - syncthing==1.22.1
  - websockify[version='>=0.10']
```
  
On the CryoHub, it times out while `Solving environment`. Any tips for getting the environment replicated?

Side note - I also tried the hub-image environment.yml and getting a similar error on my laptop:
```
Solving environment: failed

ResolvePackageNotFound:
  - qgis==3.28.2
  - syncthing==1.22.1
  - websockify[version='>=0.10']
```

**Comment by tsnow03** [comment]
Yes! I'm so sorry this hasn't been resolved yet. I'm working on this right now. I'll let you know once I have it updated!

**Comment by dfelikson** [comment]
No worries at all!

**Comment by tsnow03** [comment]
@Denis Felikson I’ve played with the environment.yml a bit. Short story is that you need to have osx>=10.15 to make this work on a local computer. Please let me know if this is a problem for you and you absolutely do not want to upgrade your osx. I can continue working to solve our environment for osx<10.15!

**Comment by dfelikson** [comment]
@tsnow03 - do you mean osx 12.15 (not 10.15)? I have 12.6 installed. I'll upgrade to 13.X eventually but NASA isn't letting us upgrade yet (they're still doing testing of that OS).

I'll try to see if I can get at least part of the environment configured so that I can get the JupyterBook up and running locally. That's all that I want to try and do right now.

Thanks for your work on this! And there's not a huge rush - this can wait until I upgrade.

**Comment by tsnow03** [comment]
That is odd. It should be 10.15 (better than Mojave) but running with mamba would tell you for sure. It probably makes more sense to just create a markdown file and commit/create a pull request and then I can test it during the review.

**Comment by dfelikson** [comment]
Thanks to @tsnow03 for pointing out that I should try `mamba`. I was actually able to get the environment configured on my laptop without error. Here's what I did:

1. `conda create --name mamba`
1. `conda activate mamba`
1. `conda install mamba -c conda-forge`
1. `mamba env create --name cryocloud --file conda/environment.yml`
1. `conda activate cryocloud`
1. `jupyter-book build book`

At this point, I got an error: `Could not import jupytext.reads (needed for MyST-NB nb_custom_formats: .py) (exception: No module named 'jupytext')`

So, I did:
1. `conda install jupytext -c conda-forge`
1. `jupyter-book build book`

And that seemed to work! I was able to open `book/_build/html/index.html` and it looked good in my browser.

I suggest to add `jupytext` to `conda/environment.yml`.
